### PR TITLE
PIM-6335: get values on variant products and product models

### PR DESCRIPTION
**Pull request by jjanvier**
Based on https://github.com/akeneo/pim-community-dev/pull/6487 that should be merged first.

## Problematic

Imagine the family `clothing`. It contains the attributes _description_, _color_, _image_, _size_ and _sku_.
This family:
- has a product `foo`
- has a family variant with
    - a `bar` root product model where we can enrich the _description_
    - a `baz` sub product model where we can enrich the _color_ and the _image_
    - a `qux` variant product where we can enrich the _size_ and the _sku_

The models, the products and the variant products do have a common method `getValues` (through the interface `Pim\Component\Catalog\Model\EntityWithValuesInterface`) which returns a `Pim\Component\Catalog\Model\ValueCollectionInterface`. The thing is to know what should return this method. We have two possibilities:

#### 1. it returns the values related to its "variation" level + the ones belonging to its ascendance

| type of entity  | entity name | `getValues` returns |
| ------------------ | --------------- | -------------------------- |
| product          | foo              | description, color, image, size, sku |
| product model  | bar  | description |
| product model | baz | description, color, image |
| variant product | qux | description, color, image, size, sku |

#### 2. it returns only the values related to its "variation" level

| type of entity  | entity name | `getValues` returns |
| ------------------ | --------------- | -------------------------- |
| product          | foo              | description, color, image, size, sku |
| product model  | bar  | description |
| product model | baz | color, image |
| variant product | qux | size, sku |

## Things to take into consideration for the choice

1. For the database storage, the method 2 is used. This is to avoid the update of the children in case a parent has been edited. A parent can have a relatively high number of children. We absolutely wanted to avoid to have to "mass edit" children each time we update a product model.

2. From a functional point of view (aka from @Delphineray), _description_, _color_, _image_, _size_ and _sku_ are all product values. Even if for the variant products, some are enriched at the root product model or sub product model level.

3. The method `getValues` is already massively used in the PIM in the following contexts:
- `ProductUniqueDataSynchronizer::synchronize` to synchronize unique data with the values of the product => here it's only for products (not models) and we want only the values of the "variation"
- `Processor\QuickExport\ProductProcessor` => all values wanted
- `VersioningBundle\Normalizer\Flat\ ProductModelNormalizer & ProductNormalizer` => normally we want only the values of the "variation", is it a problem if we have all values? to confirm with Delphine
- `ProductBuilder::addMissingProductValues` to build all missing currencies to the prices => all values wanted
- `ProductTemplateBuilder` we don't care :p 
- `CompletenessCalculator::calculate` => all values wanted
- `MetricConverter::convert` to convert metric to the desired metric => only related to `EntityWithValuesInterface` so no problem
- `CompletenessManager::getProductCompleteness` to get the completenesses of a product => all values wanted (still used?)
- `Normalizer\Standard\ProductModelNormalizer | ProductNormalizer` => **_WHAT SHOULD WE HAVE HERE_**?
- `Processor\Normalization\ProductProcessor` to be able to export medias => all values wanted

## The choice

With all considerations described above we decided that `getValues` should return all the values of an entity (not only the one that belongs to its variation level). It has 2 major advantages:
- from a functional point of view, that's the case
- this how the PIM works at the moment as there is no variant products => which means, with this PR, the CI remains blue

Maybe at some point, we'll need somewhere the only the values of the variation itself. But we didn't meet that case yet.

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added Behats                      | -
| Added integration tests           | Y
| Changelog updated                 | -
| Review and 2 GTM                  | 
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/EventSubscriber/ComputeEntityRawValuesSubscriber.php [hunk 1]
```diff
@@ -58,7 +61,53 @@ public function computeRawValues(GenericEvent $event)
...
+
+    /**
+     * Get recursively the raw values of all the parents of an entity.
+     *
+     * @param EntityWithFamilyVariantInterface $entity
+     * @param array                            $ancestryRawValues
+     *
+     * @return array
+     */
+    private function getAncestryRawValues(EntityWithFamilyVariantInterface $entity, $ancestryRawValues = []): array
```
Comment (aRn0D) [inline src/Pim/Bundle/CatalogBundle/EventSubscriber/ComputeEntityRawValuesSubscriber.php]: Missing type hint: array $ancestryRawValues

File: src/Pim/Bundle/CatalogBundle/EventSubscriber/LoadEntityWithValuesSubscriber.php [hunk 2]
```diff
@@ -77,12 +81,31 @@ public function postLoad(LifecycleEventArgs $event)
...
+
+    /**
+     * Get recursively the raw values of all the parents of an entity.
+     *
+     * @param EntityWithFamilyVariantInterface $entity
+     * @param array                            $ancestryRawValues
+     *
+     * @return array
+     */
+    private function getAncestryRawValues(EntityWithFamilyVariantInterface $entity, $ancestryRawValues = []): array
```
Comment (aRn0D) [inline src/Pim/Bundle/CatalogBundle/EventSubscriber/LoadEntityWithValuesSubscriber.php]: Missing type hint: array $ancestryRawValues

File: src/Pim/Bundle/CatalogBundle/spec/EventSubscriber/ComputeEntityRawValuesSubscriberSpec.php [hunk 3]
```diff
@@ -37,12 +42,77 @@ function it_bypasses_non_product_entities($serializer, GenericEvent $event, \std
         $serializer->normalize(Argument::cetera())->shouldNotBeCalled();
     }
 
-    function it_compute_raw_values_of_a_product($serializer, ProductInterface $product, GenericEvent $event)
-    {
+    function it_compute_raw_values_of_a_product(
```
Comment (aRn0D) [inline src/Pim/Bundle/CatalogBundle/spec/EventSubscriber/ComputeEntityRawValuesSubscriberSpec.php]: it_compute**s**

File: src/Pim/Bundle/CatalogBundle/spec/EventSubscriber/ComputeEntityRawValuesSubscriberSpec.php [hunk 4]
```diff
@@ -37,12 +42,77 @@ function it_bypasses_non_product_entities($serializer, GenericEvent $event, \std
...
+        AttributeInterface $description,
+        AttributeInterface $color,
+        AttributeInterface $image,
+        AttributeInterface $attribute,
+        GenericEvent $event
+    ) {
         $event->getSubject()->willReturn($product);
-        $product->getValues()->willReturn(['value1', 'value2']);
+        $product->getValues()->willReturn($values);
+        $values->toArray()->willReturn([$descriptionValue, $colorValue, $imageValue, $value1, $value2]);
```
Comment (aRn0D) [inline src/Pim/Bundle/CatalogBundle/spec/EventSubscriber/ComputeEntityRawValuesSubscriberSpec.php]: Don't understand where do you call this method?

File: src/Pim/Bundle/CatalogBundle/tests/integration/Doctrine/Common/Saver/ProductSaverIntegration.php [hunk 5]
```diff
@@ -45,7 +71,7 @@ protected function getConfiguration()
      *
      * @return ProductInterface
      */
-    private function createProduct($productIdentifier, $familyCode)
+    private function createProduct(string $productIdentifier, string $familyCode)
```
Comment (aRn0D) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Doctrine/Common/Saver/ProductSaverIntegration.php]: You forgot to add the missing return type for this method and the other one :trollface:

File: src/Pim/Bundle/CatalogBundle/EventSubscriber/ComputeEntityRawValuesSubscriber.php (hunk 1 above)
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/EventSubscriber/ComputeEntityRawValuesSubscriber.php]: done

File: src/Pim/Bundle/CatalogBundle/EventSubscriber/LoadEntityWithValuesSubscriber.php (hunk 2 above)
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/EventSubscriber/LoadEntityWithValuesSubscriber.php]: done

File: src/Pim/Bundle/CatalogBundle/spec/EventSubscriber/ComputeEntityRawValuesSubscriberSpec.php (hunk 3 above)
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/spec/EventSubscriber/ComputeEntityRawValuesSubscriberSpec.php]: done

File: src/Pim/Bundle/CatalogBundle/spec/EventSubscriber/ComputeEntityRawValuesSubscriberSpec.php (hunk 4 above)
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/spec/EventSubscriber/ComputeEntityRawValuesSubscriberSpec.php]: in the construction of the collection itself

File: src/Pim/Bundle/CatalogBundle/EventSubscriber/ComputeEntityRawValuesSubscriber.php [hunk 6]
```diff
@@ -58,7 +61,53 @@ public function computeRawValues(GenericEvent $event)
...
+            if (in_array($value->getAttribute()->getCode(), $ancestryAttributeCodes)) {
+                $values->remove($value);
+            }
+        }
+
+        return $values;
+    }
+
+    /**
+     * Get recursively the raw values of all the parents of an entity.
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/EventSubscriber/ComputeEntityRawValuesSubscriber.php]: done

File: src/Pim/Bundle/CatalogBundle/EventSubscriber/ComputeEntityRawValuesSubscriber.php [hunk 7]
```diff
@@ -58,7 +61,53 @@ public function computeRawValues(GenericEvent $event)
...
+    /**
+     * Remove all the parent values of an entity.
+     *
+     * @param EntityWithFamilyVariantInterface $entity
+     *
+     * @return ValueCollectionInterface
+     */
+    private function removeAncestryValues(EntityWithFamilyVariantInterface $entity): ValueCollectionInterface
+    {
+        // copy the values here so that we don't touch the initial values collection
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/EventSubscriber/ComputeEntityRawValuesSubscriber.php]: done

File: src/Pim/Bundle/CatalogBundle/EventSubscriber/LoadEntityWithValuesSubscriber.php [hunk 8]
```diff
@@ -77,12 +81,31 @@ public function postLoad(LifecycleEventArgs $event)
...
     {
         if (null === $this->valueCollectionFactory) {
             $this->valueCollectionFactory = $this->container->get('pim_catalog.factory.value_collection');
         }
 
         return $this->valueCollectionFactory;
     }
+
+    /**
+     * Get recursively the raw values of all the parents of an entity.
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/EventSubscriber/LoadEntityWithValuesSubscriber.php]: done

**Pull request merged**